### PR TITLE
hybridize rnn and add model graph

### PR DESCRIPTION
**Pull request by yifeim**
## Description ##
Add hybridize option and visualize the network in mxnet>=1.3.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Add hybridize option. Hybridize on RNN layers is newly available in mxnet>=1.3.
- [x] Add model export to mxboard. The end result is visualized via tensorboard and added to README.

## Comments ##
- The change is back-ward compatible. Especially, hybridize_optional would be an empty list in mxnet<1.3.

### CODE REVIEW

**Comment by kalyc** [comment]
@mxnet-label-bot add [pr-work-in-progress]

**Comment by stu1130** [comment]
@mxnet-label-bot update [pr-awaiting-review, gluon]

File: example/gluon/word_language_model/README.md [hunk 1]
```diff
@@ -28,7 +28,9 @@ python train.py --cuda --tied --nhid 650 --emsize 650 --epochs 40  --dropout 0.5
 ```
 python train.py --cuda --tied --nhid 1500 --emsize 1500 --epochs 60  --dropout 0.65     # Test ppl of 88.42
 ```
-
+```
+python train.py --export-only # hybridize and export model graph (requires mxboard)
```
Comment (vandanavk) [inline example/gluon/word_language_model/README.md]: Please add instruction to install mxboard

File: example/gluon/word_language_model/train.py [hunk 2]
```diff
@@ -58,6 +58,14 @@
...
 parser.add_argument('--gcthreshold', type=float, default=0.5,
                     help='threshold for 2bit gradient compression')
+parser.add_argument('--hybridize', action='store_true',
+                    help='whether to hybridize in mxnet>=1.3')
+parser.add_argument('--static-alloc', action='store_true',
+                    help='whether to use static-alloc hybridize in mxnet>=1.3')
+parser.add_argument('--static-shape', action='store_true',
+                    help='whether to use static-shape hybridize in mxnet>=1.3')
+parser.add_argument('--export-only', action='store_true',
+                    help='export a symbol graph and exit')
```
Comment (vandanavk) [inline example/gluon/word_language_model/train.py]: Please mention the default values for all of these arguments

File: example/gluon/word_language_model/README.md (hunk 1 above)
Comment (yifeim) [reply example/gluon/word_language_model/README.md]: added link and made it outside of the main codes.

File: example/gluon/word_language_model/train.py (hunk 2 above)
Comment (yifeim) [reply example/gluon/word_language_model/train.py]: Added defaults as directed.

**Comment by vandanavk** [comment]
@roywei for review

**Comment by yifeim** [comment]
Is this related to my PR?
http://jenkins.mxnet-ci.amazon-ml.com/blue/organizations/jenkins/incubator-mxnet/detail/PR-13244/5/pipeline
```
[INFO] Setting module np/mx/python random seeds, use MXNET_MODULE_SEED=1716360676 to reproduce.
test_gluon_model_zoo_gpu.test_inference ... [02:33:54] src/io/iter_image_recordio_2.cc:172: ImageRecordIOParser2: data/val-5k-256.rec, use 1 threads for decoding..
[02:33:57] src/operator/nn/mkldnn/mkldnn_base.cc:74: Allocate 32112640 bytes with malloc directly
[02:34:03] src/io/iter_image_recordio_2.cc:172: ImageRecordIOParser2: data/val-5k-256.rec, use 1 threads for decoding..
Sending interrupt signal to process
build.py: 2018-11-27 02:34:52,714 Signal 15 received, cleaning up...
build.py: 2018-11-27 02:34:52,714 Cleaning up containers
Terminated
build.py: 2018-11-27 02:34:57,229 ☠: stopped container 3c9b21b903de
build.py: 2018-11-27 02:34:57,290 🚽: removed container 3c9b21b903de
build.py: 2018-11-27 02:34:57,291 Cleaning up containers finished.
build.py: 2018-11-27 02:34:57,291 done. Exiting with error.
script returned exit code 143
```

**Comment by roywei** [comment]
@yifeim Thanks for the contribution. Is there any performance improvement after hybridize? Btw, you can trigger CI again by rebase or push an empty commit.

**Comment by yifeim** [comment]
I did not see any performance improvement after hybridization. My guess is
that the computations are already elementary and asynchronous.

**Comment by yifeim** [comment]
1. As noted in both README and help strings, hybridizing an RNN is not
supported in mxnet<=1.2.

2. Export is the most obvious way for me to dump a network, unless you
recommend a different approach that I am not aware of?

**Comment by yifeim** [comment]
*@sandeep-krishnamurthy*

If I understand correctly, perhaps I should also emphasize that there are
models that are never hybridized for good reasons. Hybridize should always
be “an option”, for two reasons:

1. For some algorithms, e.g., RNN models, hybridize offers no performance
differences.

2. Hybridize turns a gluon model into a symbol model and hence it strips
away all dynamic computation capabilities, such as if/else, for loop, and
array indexing. While mxnet 1.3 offers some dynamic graph capabilities,
they are still very limited. For example, if/else condition requires the
outputs to always have the same dimensions, which is sometimes impractical.

**Comment by yifeim** [comment]
Ok I agree.

**Comment by yifeim** [comment]
Can you make the changes? Thanks!

**Comment by anirudhacharya** [comment]
@sandeep-krishnamurthy can you please review/merge this PR?

**Comment by sandeep-krishnamurthy** [comment]
@yifeim - Sorry for delayed response.
Rather than calling it "export_only", can we call it "export-model"? So it will be more clear for users?

**Pull request merged**